### PR TITLE
[spring-boot-book-store-creteria-api-hw] implemented search by author, title and price

### PR DESCRIPTION
**Pull request by Andriy-Hrynevetskyi**
- added search endpoint and method at controller layer;
- added dto for search params;
- added specificationBuilder;
- added specificationBuilderProvider;
- added specificationBuilderProviderManager;
- added specifications;

### CODE REVIEW

**Comment by Andriy-Hrynevetskyi** [comment]
I have  doubts about serching by price. I couldn't realize it using some kind of "between" (between min & max price).

File: src/main/java/bookstore/repository/book/spec/TitleSpecificationProvider.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import bookstore.repository.SpecificationProvider;
+import java.util.Arrays;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "title";
```
Comment (kshuryhin) [inline src/main/java/bookstore/repository/book/spec/TitleSpecificationProvider.java]: Good candidate to be constant I guess

File: src/main/java/bookstore/repository/book/spec/PriceSpecificationProvider.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import bookstore.repository.SpecificationProvider;
+import java.util.Arrays;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class PriceSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "price";
```
Comment (kshuryhin) [inline src/main/java/bookstore/repository/book/spec/PriceSpecificationProvider.java]: Good candidate to be constant I guess

File: src/main/java/bookstore/repository/book/BookSpecificationBuilder.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            spec = spec.and(bookSpecificationProviderManager.getSpecificationProvider("author")
+                    .getSpecification(searchParameters.authors()));
+        }
+        if (searchParameters.titles() != null && searchParameters.titles().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager.getSpecificationProvider("title")
+                    .getSpecification(searchParameters.titles()));
+        }
+        if (searchParameters.prices() != null && searchParameters.prices().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager.getSpecificationProvider("price")
+                    .getSpecification(searchParameters.prices()));
```
Comment (kshuryhin) [inline src/main/java/bookstore/repository/book/BookSpecificationBuilder.java:29]: I guess we can split it into several methods for better code readability. Also, we can use Optional for the null checks and streamline the code

File: src/main/java/bookstore/repository/book/AuthorSpecificationProvider.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package bookstore.repository.book;
+
+import bookstore.model.Book;
+import bookstore.repository.SpecificationProvider;
+import java.util.Arrays;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
```
Comment (kshuryhin) [inline src/main/java/bookstore/repository/book/AuthorSpecificationProvider.java]: I think this one should be moved to the **spec** package, what do you think?

**Review by kshuryhin — CHANGES_REQUESTED** [review]
Good job, but there are a few points to consider

File: src/main/java/bookstore/repository/book/BookSpecificationBuilder.java (hunk 3 above)
Comment (Andriy-Hrynevetskyi) [reply src/main/java/bookstore/repository/book/BookSpecificationBuilder.java:29]: Sorry, but i think that i am not able to do this without someone's help(((

Comment (vovasalyha) [reply src/main/java/bookstore/repository/book/BookSpecificationBuilder.java:29]: At least you can extract repetitive checks `searchParameters.prices() != null && searchParameters.prices().length > 0` into separate private method like `isParamValid` etc.

**Review by vovasalyha — APPROVED** [review]
LGTM

**Pull request merged**